### DISCUSSION TITLE
support for chrome apps (extentions) OAuth 2

### DISCUSSION THREAD

**Discussion by whateverneveranywhere** [pocketbase/pocketbase#1899] [Q&A] [answered]
Hey there!

first of all. thanks for the amazing product. been a rocket launcher in terms of how much boost it added to my side projects!

one limitation I noticed recently, 
using manifest v3 and the latest chrome extension APIs as of today, 
the flow is done using `chrome.identity.getAuthToken({ interactive: true, })` an in-browser popup to log in with a google account, and if you check google's cloud console, it is a requirement to specifically use "chrome app" to get a client ID for OAuth 2 if the project is a chrome extension.

![image](https://user-images.githubusercontent.com/29921792/220698237-e5e22acc-cffb-4f0d-8340-059b42bc5df4.png)

however, using the method offered in javascript's SDK using ` pb.collection('users').authWithOAuth2` there is multiple stuff that is required by design but isn't compatible with this flow. 

![image](https://user-images.githubusercontent.com/29921792/220700304-666a7154-084c-4fa2-82f3-8235acc7d626.png)

for example, in the panel, when using a chrome app, there is no need for a `client secret` but it's a required field and later on, the need for a callback URL for when the process was successful, is nothing that should be forced since this flow may fully happen just in the browser and in the extension's popup. 

![image](https://user-images.githubusercontent.com/29921792/220700526-3ca2ec1b-23ef-48f6-bfac-c23050ada193.png)

I'm not sure if there's a workaround for this at this moment, but would really appreciate it if this could be considered as a feature sometime in the future.

Again, thank you so much for the awesome project <3

**Comment by ganigeorgiev** [comment, marked as answer]
This is too platform and provider specific and I don't think we can implement it as part of the generic OAuth2 flow.

Similar to the OAuth2 in Android (https://github.com/pocketbase/pocketbase/discussions/442#discussioncomment-4160820), one workaround could be to select "Web application" as application type and implement it as in a regular web app, although I'm not sure if there are any requirements or policy restrictions related to this when publishing the Chrome app to the web store.

The OAuth2 flow will be further simplified in the near future as part of #55 (_if you haven't already, you can check the current planned tasks and their priority in https://github.com/orgs/pocketbase/projects/2_).

**Reply by ganigeorgiev** [reply]
Another workaround (in addition to the "Web application") could be also to use the [`chrome.identity.launchWebAuthFlow`](https://developer.chrome.com/docs/extensions/reference/identity/#method-launchWebAuthFlow):

> **This method enables auth flows with non-Google identity providers by launching a web view and navigating it to the first URL in the provider's auth flow.** When the provider redirects to a URL matching the pattern `https://<app-id>.chromiumapp.org/*`, the window will close, and the final redirect URL will be passed to the callback function. 

In general, the idea is to treat Google as generic OAuth2 identity provider.

**Reply by whateverneveranywhere** [reply]
> This is too platform and provider specific and I don't think we can implement it as part of the generic OAuth2 flow.
> 
> Similar to the OAuth2 in Android ([#442 (comment)](https://github.com/pocketbase/pocketbase/discussions/442#discussioncomment-4160820)), one workaround could be to select "Web application" as application type and implement it as in a regular web app, although I'm not sure if there are any requirements or policy restrictions related to this when publishing the Chrome app to the web store.
> 
> The OAuth2 flow will be further simplified in the near future as part of #55 (_if you haven't already, you can check the current planned tasks and their priority in https://github.com/orgs/pocketbase/projects/2_).

there are actually some policy restrictions by google to authenticate a plugin login using a separate new tab. :(
the workaround suggested at (https://github.com/pocketbase/pocketbase/discussions/442#discussioncomment-4160820), wasn't helpful for my case. 

I think I didn't express the part that I'm having issues with correctly, so, in my specific case I am able to get the user logged in using `chrome.identity.getAuthToken` without any issues and it grants me the token and it's valid to get all the data defined in the manifest scope. 

I want to create a new `user` record in pocketbase and since this has been an OAuth token, I want to skip the required `password` validation rules.

Any possible workaround that'd work right now or probably any future updates that'd allow OAuth handling like this?

**Reply by ganigeorgiev** [reply]
@whateverneveranywhere Please read the previous comment and https://developer.chrome.com/docs/extensions/reference/identity/#method-launchWebAuthFlow.

**Reply by whateverneveranywhere** [reply]
@ganigeorgiev  
trying that approach it'll show policy limitations rules:

EDIT: previous screen shot was not related to this, replaced, sorry :D 

![image](https://user-images.githubusercontent.com/29921792/220934187-d2c50374-a34f-4a48-a9cb-3fc66a24e1d9.png)

**Reply by ganigeorgiev** [reply]
Could you provide a code sample how you are making the request?

Are you using the `authUrl` generated from `listAuthMethods()` call?

**Reply by ganigeorgiev** [reply]
Also make sure that your Google OAuth2 app is properly configured (aka. your user exist in the "tests users" list if it is not published yet).

**Reply by whateverneveranywhere** [reply]
sure, here's the component I'm making this request from (React v18, manifest v3, chrome extension):

![image](https://user-images.githubusercontent.com/29921792/220937137-2c39e669-91c1-4a6e-b920-82c277a1b201.png)

as suggested in the docs, saved the available method to local storage first,
then on click of a button trying to use [chrome.identity.launchWebAuthFlow](https://developer.chrome.com/docs/extensions/reference/identity/#method-launchWebAuthFlow):

![image](https://user-images.githubusercontent.com/29921792/220938773-467c9093-8275-479c-850b-fee7ed9ac471.png)

then the popup opens with that policy error. :(

**Reply by whateverneveranywhere** [reply]
> Also make sure that your Google OAuth2 app is properly configured (aka. your user exist in the "tests users" list if it is not published yet).

checked already, the app is published, using `web app` type instead of `chrome app`, wasn't an issue with that.

**Reply by ganigeorgiev** [reply]
I don't think the policy error is related to the Chrome app/extension apis.

There is no `googleAuth.redirectUri` field in the default `listAuthMethods()` response.
Make sure that it is not empty and matches the one configured in your Google OAuth2 app console.

I suspect that the OAuth2 app is not properly configured. To confirm that this is the case, try to reproduce the error using the plain js web example as shown in https://pocketbase.io/docs/authentication/#web-oauth2-integration (aka. create the 2 static html files and place them inside your `pb_public` and configure the redirect url in the Google console).

If that's the case, make sure that:
- your app has properly configured redirect address
- your user is listed in the "Test users" section in the "Google console > OAuth consent screen"

**Reply by whateverneveranywhere** [reply]
wrong screenshot again, sorry :D  been trying so many approaches for the past few days, a bit all over the place even trying to explain the case. :D 

`const redirectUri = chrome.runtime.getManifest().oauth2.redirect_url;`

I'm getting the redirect URL from my manifest file, already registered so google won't have issues with security, that is in this format:

`"redirect_url": https://<client-id>.chromiumapp.org`

**Reply by whateverneveranywhere** [reply]
> I don't think the policy error is related to the Chrome app/extension apis.
> 
> There is no `googleAuth.redirectUri` field in the default `listAuthMethods()` response. Make sure that it is not empty and matches the one configured in your Google OAuth2 app console.
> 
> I suspect that the OAuth2 app is not properly configured. To confirm that this is the case, try to reproduce the error using the plain js web example as shown in https://pocketbase.io/docs/authentication/#web-oauth2-integration (aka. create the 2 static html files and place them inside your `pb_public` and configure the redirect url in the Google console).
> 
> If that's the case, make sure that:
> 
> * your app has properly configured redirect address
> * your user is listed in the "Test users" section in the "Google console > OAuth consent screen"

Thank you so much! 
will look a bit more into it now. I just noticed the static HTML in `pb_public`. let me try that now and will update the thread with my findings again <3

**Reply by whateverneveranywhere** [reply]
@ganigeorgiev hello again!
totally sorry about the previous issue. the mismatching URL was totally my fault, not setting the redirect path correctly on the dev console and not reading documents well enough and thanks for the patient explanations and solution suggestions, really helpful. <3 

![image](https://user-images.githubusercontent.com/29921792/220986047-cb4f0825-4d21-4be3-93d8-d223e97bd9d4.png)

I got it working but the issue is now that [chrome.identity.launchWebAuthFlow](https://developer.chrome.com/docs/extensions/reference/identity/#method-launchWebAuthFlow) will open a new popup that doesn't have access to the already opened extension's popup.
Couldn't find a safe workaround for it either and it's not the same as the case with android I think, I checked that solution. 

any suggestions to somehow allow this new popup to be aware of the extension's popup localstorage data?

**Reply by ganigeorgiev** [reply]
Based on the `launchWebAuthFlow` documentation the auth window should autoclose if your redirect uri matches the chromiumapp pattern and you should receive the final redirect url in its callback function:

> When the provider redirects to a URL matching the pattern `https://<app-id>.chromiumapp.org/*`, **the window will close, and the final redirect URL will be passed to the callback function.**

If that's not the behavior you are experiencing, please provide a more detailed code sample with your current implementation and I'll try to test it.

**Reply by whateverneveranywhere** [reply]
I think I've totally got it wrong then. 
see, on click of my button I am running this code:

![image](https://user-images.githubusercontent.com/29921792/220992824-6b760524-a4da-402f-a55b-b7ef54653673.png)

I've put the `pb` live on a Debian 11 server with https flag served and the domain you see in the picture is not the real domain I'm using.
I've also created the `redirect.html` in `pb_public` and exactly copy pasted the content from what was in the documentation (and change URLs accordingly).

![image](https://user-images.githubusercontent.com/29921792/220996148-773b9556-993a-4bce-ac6f-97a79481ca63.png)

it won't close the popup and get to the extension with the callback because, in the `redirect.html` line 10, there's no "provider" in the local storage of this newly created popup.

also, my manifest v3 file looks something like this:

`
  "oauth2": {
    "client_id": "client-id",
    "scopes": [
      "https://www.googleapis.com/auth/userinfo.profile",
      "https://www.googleapis.com/auth/userinfo.email",
    ],
    "redirect_url": "https://livewebsite.test/redirect.html"
  },
`


in the documentation, since it's assumed that the href will be opened in the current window, it will be able to access local storage without issues, but in my case after choosing the google account to login within the `launchWebAuthFlow `'s auth, it'll be redirected to `https://livewebsite.test/redirect.html` and obviously, it won't have access to local storage of the plugin's popup there since it's considered a new window totally.

**Reply by ganigeorgiev** [reply]
When using the `launchWebAuthFlow` method the redirect should have the format `https://<app-id>.chromiumapp.org/*` if you want to handle it inside its callback (_I've previously suggested to test with the static html example just to confirm that the error in https://github.com/pocketbase/pocketbase/discussions/1899#discussioncomment-5088985 was not related to the extension and was a result of misconfigured OAuth2 app_).

---

For example, to generate the url `https://<app-id>.chromiumapp.org/oauth2`, you can use [`chrome.identity.getRedirectURL("oauth2")`](https://developer.chrome.com/docs/extensions/reference/identity/#method-getRedirectURL).

Please read carefully the documentation in https://developer.chrome.com/docs/extensions/reference/identity/#method-launchWebAuthFlow and the other helper methods.

**Reply by whateverneveranywhere** [reply]
@ganigeorgiev okay, changed the app and now it looks like this in the dev console: 

1- dev console:

![image](https://user-images.githubusercontent.com/29921792/221023803-3331b702-a654-463e-84e3-a3952471caa9.png)

2- manifest v3:

`
  "oauth2": {
    "client_id": "code-something.apps.googleusercontent.com",
    "scopes": [
      "https://www.googleapis.com/auth/userinfo.profile",
    ],
    "redirect_url": "https://something.chromiumapp.org/login"
  },
`
3- code to call the `launchWebAuthFlow`:

![image](https://user-images.githubusercontent.com/29921792/221024446-76b7254e-7713-4373-b53e-c789d8d90b17.png)

![image](https://user-images.githubusercontent.com/29921792/221025409-fcc26dba-401e-4738-b93e-c11beaff7506.png)

the reason I've changed the authUrl to include the `code_verifier` was because the popup asked for it first, and when given the `code_verifier` it showed the policy error below with a status code of 400:

![image](https://user-images.githubusercontent.com/29921792/221025720-f8685ec6-cd11-4203-948f-01e7c81cad60.png)

**Comment by ganigeorgiev** [comment]
I'll answer as another reply, sine the topic started to become hard to follow.

I don't understand from where the confusion come and it seems that we are going in circles.

**Once again, please read carefully the documentation of the [`launchWebAuthFlow`](https://developer.chrome.com/docs/extensions/reference/identity/#method-launchWebAuthFlow) method.**

As explained in the ["Web OAuth2 integration" example](https://pocketbase.io/docs/authentication/#web-oauth2-integration), to integrate with any OAuth2 provider you need 2 things:

1. Somewhere to show the "Login with ..." links.
   This is where you call `listAuthMethods()` to get the prebuilt provider `authUrl` and concatenate it with your redirect url.

2. Somewhere to handle the OAuth2 redirect.
   In the above example, we are using a separate static html page, but in the extension case you have to process the result of `launchWebAuthFlow`(either as callback argument or the returned Promise).

Here is a simplified code sample (_I haven't tested it_):

```js
const redirectUrl = chrome.identity.getRedirectURL("oauth2");

const authMethods = await pb.collection('users').listAuthMethods();

const googleAuth = authMethods.authProviders.find((provider) => provider.name === "google");

const responseUrl = await chrome.identity.launchWebAuthFlow({
    url: googleAuth.authUrl + redirectUrl,
    interactive: true,
})

const responseParams = (new URL(responseUrl)).searchParams;

// (this check is not required for this particular use case)
if (googleAuth.state !== responseParams.get("state")) {
    throw "State parameters don't match.";
}

// sign-in/sign-up in PocketBase
await pb.collection("users").authWithOAuth2(
    googleAuth.name,
    responseParams.get("code"),
    googleAuth.codeVerifier,
    redirectUrl,
    // pass optional user create data...
    {
        emailVisibility: false,
    }
);

console.log("Success:", pb.authStore.model.id);
```

**Reply by saphal1998** [reply]
This doesn't seem to work for me, I still seem to be getting a `redirect_uri_mismatch` when trying to login via Google in the Chrome Extension.

<img width="395" alt="Screenshot 2023-07-30 at 1 34 39 AM" src="https://github.com/pocketbase/pocketbase/assets/31125345/07ce8d79-a5ac-4e12-a844-2ddc06afd92f">

There isn't an option to set a callback URL for OAuth Credentials generated for Chrome Applications, but I assume that the default redirect URL is going to be used, as per [Basics, Authentication](https://pocketbase.io/docs/authentication/) 
```
This method handles everything within a single call without having to define custom redirects, deeplinks or even page reload.

When creating your OAuth2 app, for a callback/redirect URL you have to use https://yourdomain.com/api/oauth2-redirect.
```

So, I used this as my redirect URL
```
  const redirectUrl = chrome.identity.getRedirectURL("api/oauth2-redirect")
```

But this doesn't seem to work. 

Here's my complete code

```
async function startAuth() {
  const authMethods = await pb.collection('users').listAuthMethods()
  const googleAuth = authMethods.authProviders.find((provider) => provider.name === "google");
  console.log("🚀 ~ file: background.ts:28 ~ startAuth ~ googleAuth:", googleAuth)
  const redirectUrl = chrome.identity.getRedirectURL("api/oauth2-redirect")
  const responseUrl = await chrome.identity.launchWebAuthFlow({
    url: googleAuth.authUrl + redirectUrl,
    interactive: true,
  })

// This seems to work if I just get the token using gerAuthToken
  //  const { token } = chrome.identity.getAuthToken({ interactive: true })
  // const userInfoStream = await fetch("https://www.googleapis.com/oauth2/v1/userinfo?alt=json", {
  //   headers: {
  //     "Authorization": `Bearer ${token}`
  //   }
  // })
  // const userInfo = await userInfoStream.json()
  // console.log("🚀 ~ file: background.ts:33 ~ startAuth ~ userInfo:", userInfo)
  // const user = {
  //   id: userInfo.id,
  //   email: userInfo.email,
  //   name: userInfo.name,
  //   avatar: userInfo.picture,
  //   username: userInfo.email
  // }
  // console.log("🚀 ~ file: background.ts:40 ~ startAuth ~ user:", user)

  return true;
}
```


I would really appreciate if you could tell me how to get an OAuth flow working in a Chrome Extension. My intention is to use Google's OAuth to populate the user table I have created in my pocketbase instance.

**Reply by ganigeorgiev** [reply]
Make sure that the generated `redirectUrl` matches the one you have configured in your Google OAuth2 app.

An alternative approach could be also to use the ["all-in-one" flow](https://pocketbase.io/docs/authentication/#oauth2-integration).

```js
const authData = await pb.collection('users').authWithOAuth2({
    provider: 'google',
    // optional custom urlCallback (by default opens a new window but this may not be allowed in extensions)
    urlCallback: (url) => {
        // open the sign-in url in a new tab or window...
    }
});
```

**Reply by ganigeorgiev** [reply]
> There isn't an option to set a callback URL for OAuth Credentials generated for Chrome Applications

As mentioned in https://github.com/pocketbase/pocketbase/discussions/1899#discussioncomment-5079634, you can try to select "Web application" as application type and implement it as in a regular web app.

**Reply by saphal1998** [reply]
I don't think there's an option to set the redirect url for Chrome Applications, I have tried the web application approach as well as using the all-in-one approach too, but both of them seem to give me a `redirect_uri_mismatch` error.

> Make sure that the generated `redirectUrl` matches the one you have configured in your Google OAuth2 app.
> 
> An alternative approach could be also to use the ["all-in-one" flow](https://pocketbase.io/docs/authentication/#oauth2-integration).
> 
> ```js
> const authData = await pb.collection('users').authWithOAuth2({
>     provider: 'google',
>     // optional custom urlCallback (by default opens a new window but this may not be allowed in extensions)
>     urlCallback: (url) => {
>         // open the sign-in url in a new tab or window...
>     }
> });
> ```

**Reply by ganigeorgiev** [reply]
@saphal1998 To be able to specify a redirect url you need to select **"Web application"** as application type when creating the Google OAuth2 app.

**Reply by saphal1998** [reply]
This is my popup.html (I've got it working with Svelte). I'm hosting pocketbase on `https://<self_hosted>.hop.sh`.

```ts
<script lang="ts">
  import PocketBase from "pocketbase";
  const pb = new PocketBase("https://<self_hosted>.hop.sh");

  async function handleLoginAuth() {
    const authData = await pb
      .collection("users")
      .authWithOAuth2({ provider: "google" });
  }
</script>

<main class="extension">
  {#if !pb.authStore.isValid}
    <h2>
      You are not logged in, please login to continue using this extension
    </h2>
    <button class="login-button" on:click={handleLoginAuth}>Login</button>
  {:else}
    <h2>
      Hi {pb.authStore.model.email}, you can now use this extension to track
      your scores
    </h2>
  {/if}
</main>
```

Here's my manifest file
```json
  "oauth2": {
    "client_id": "<client_id_from_gcp>",
    "scopes": [
      "email",
      "profile",
      "openid"
    ]
  },
```

I've also made sure to add a key in the manifest file to ensure my extension ID does not change between rebuilds as specified [here](https://developer.chrome.com/docs/extensions/mv3/tut_oauth/#keep-consistent-id)

I've also ensured that I have put the right client ID and client secret (now that I'm using the WebApp option) from GCP.
<img width="827" alt="Screenshot 2023-07-30 at 12 53 31 PM" src="https://github.com/pocketbase/pocketbase/assets/31125345/215ecfa1-7ea3-49dd-82ec-127e1f482938">

My authorized Javascript origins on my GCP are
```https://<extension_id>.chromium.app``` where extension id is the same as I generated on the Chrome Extension Developer Console.

My authorized Javascript redirects on my GCP are
```https://<extension_id>.chromium.app/api/oauth2-redirect``` where extension id is the same as I generated on the Chrome Extension Developer Console.



I think the issue is with the redirect URI if I'm being honest, but I'm not sure what to specify it as. Do you have any pointers about that?



I tried to set the Redirect URI to `https://<self_hosted>.hop.sh/api/oauth2-redirect` but that seems to prompt me for my gmail and then gives me a window saying
`{"code":404,"message":"Not Found.","data":{}}` (The `window.location.href` for that window is `https://<self_hosted>.hop.sh/api/oauth2-redirect?state=lPZEulkGIIR9QfaQJbitqVfByIXWGzGvIsiVK3n2&code=4%2F0AZEOvhXk5DF9TnAmchLyL2YABZqox85oH6InpkqOK4dDz4rdwrqPaf5W4uO8pfgjppHMVg&scope=email+profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+openid&authuser=0&prompt=consent`

<img width="1461" alt="Screenshot 2023-07-30 at 1 03 43 PM" src="https://github.com/pocketbase/pocketbase/assets/31125345/ff294cff-62db-4ac4-8ebc-8a9a655ec3f3">



And now, I'm out of ideas :/ 

Does this help maybe narrow down the issue?

**Reply by ganigeorgiev** [reply]
@saphal1998 You've written "_I've got it working with Svelte_" but at the same time you claim that it doesn't work with the extension setup. Why? Where is the difference? Did you setup the Google OAuth2 app as "Web application"?

Sorry but I think we are going in circles and I'm not sure that I'll be able to help you as I don't understand your setup and what you are trying to do.

To help you with the debugging, you can try starting PocketBase `--debug` and check if there any additional errors printed in the console.

**Reply by saphal1998** [reply]
I meant that my extension's UI is setup with Svelte.  I'm using "Web Application".

But I understand, thank you so much for your efforts!

**Reply by ganigeorgiev** [reply]
Here is where the "Web application" option can be selected:

![oauth2_app_dropdown](https://i.imgur.com/pulJvxu.png).

If the above doesn't help, I'm out of ideas.

**Reply by saphal1998** [reply]
I have selected this, but I really appreciate you going above and beyond to resolve this. Thank you so much! 

I'll post a solution if I find one.

**Reply by ganigeorgiev** [reply]
@saphal1998 If you had selected "Web application" for your OAuth2 app and have setup the correct redirect URIs in:

![redirect_uris](https://i.imgur.com/Ya0VYAS.png)

then the ["all-in-one" OAuth2 flow](https://pocketbase.io/docs/authentication/#oauth2-integration) should work fine.

Note that the redirect URI in your case should be `https://<self_hosted>.hop.sh/api/oauth2-redirect`.

**Reply by saphal1998** [reply]
I am able to get the callback in my urlCallback (`console.log(callback)`)

```
  async function handleLoginAuth() {
    const authData = await pb.collection("users").authWithOAuth2({
      provider: "google",
      urlCallback: async (...params) => {
        const callback = params[0];
        console.log(callback);
      }
    });
  }
```

But idk what to do after that, I don't have access to any of the fields specified in my scope (i.e. email, name of the user etc..).

For completeness, here's the output of the `console.log(callback)`.

`https://accounts.google.com/o/oauth2/auth?client_id=<client_id_from_gcp>.apps.googleusercontent.com&code_challenge=AGZZmscQYBPjyjL23H5Ceax_0syszbMOMCVFnAhitiM&code_challenge_method=S256&response_type=code&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email&state=0BUB99ARKe2YH4WRgulYkRS3Ho1V7v9BHJCOFD46&redirect_uri=https%3A%2F%2F<self_hosted>.hop.sh%2Fapi%2Foauth2-redirect`

What do I do from here to populate the user's information in the database and have the session data be present in pb.authStore?

**Reply by saphal1998** [reply]
As far as I can tell from your example, this should have `code` in its search params? But it doesn't in this callback url?

**Reply by ganigeorgiev** [reply]
You need to open this url in a new window/tab. The default action if you don't overwrite the `urlCallback` will try to open the url automatically in a new browser window.
Once you complete the provider sign-in process, it will automatically create the PocketBase user for you and will return the success response as result of the `authWithOAuth2` Promise.

Also note that `urlCallback` receives always only 1 argument (the sign-in url) and the `...params` part is not needed, just replace it with `url` (_if you need to specify custom `urlCallback` in the first place; this part I haven't tested and I'm not sure whether Chrome apps or extensions allow it or have a dedicated method for such use cases_).

**Reply by saphal1998** [reply]
> Once you complete the provider sign-in process, it will automatically create the PocketBase user for you and will return the success response as result of the `authWithOAuth2` Promise.

I tried to remove the urlCallback and well, it does open a popup asking me to login. But after I select my email, I see this screen.

<img width="790" alt="Screenshot 2023-07-30 at 2 04 45 PM" src="https://github.com/pocketbase/pocketbase/assets/31125345/5e625b60-f8a2-4efe-9be8-75ded933243d">

My hypothesis is that this comes up because the user doesn't exist? The corresponding request log in pocketbase is this. 

<img width="662" alt="Screenshot 2023-07-30 at 2 14 15 PM" src="https://github.com/pocketbase/pocketbase/assets/31125345/c9de0bff-1f85-4475-acd2-3c91e128f140">

Any thoughts?

**Reply by ganigeorgiev** [reply]
When using the "all-in-one" OAuth2 flow the page that initialized the OAuth2 call needs to be running in the background.

If it was closed, then the Google redirect will fail because the persistent connection between the client and the PocketBase server was lost.

I really don't know how to help you further. If you think that this is a PocketBase issue, please create a **minimal** reproducible repo and tomorrow I'll try to investigate it further.

**Reply by saphal1998** [reply]
https://github.com/saphal1998/pocketbase-oauth2-example 

Here's a minimal reproducible repo for what I'm trying to do, I've also included the Dockerfile that I used to deploy my pocketbase instance. 

Please let me know if anything else is needed for you to get more context on this.

**Reply by ganigeorgiev** [reply]
@saphal1998 I should have clarified it earlier, but looking at your [Dockerfile](https://github.com/saphal1998/pocketbase-oauth2-example/blob/master/Dockerfile#L4) it seems that you are using an older PocketBase v0.14.0 (unless you overwrite somewhere else the `PB_VERSION` build argument) **which doesn't support the "all-in-one" OAuth2 flow** (that would also explain the 404 error because the oauth2 redirect route doesn't exist).

Please try again with the [latest PocketBase v0.17.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.17.0) (the "all-in-one" OAuth2 flow was introduced with PocketBase v0.15.0).

**Reply by saphal1998** [reply]
YES! That works! Thank you so much! 

Amazing how simple it was, thank you so much once again for being so patient
